### PR TITLE
Implement initial client calls on the Cloud API facade.

### PR DESCRIPTION
**Pull request by frankban**
Also properly establish both model and controller connections in sandbox mode.

### CODE REVIEW

**Comment by frankban** [comment]
QA:
```javascript
// In sandbox mode.

app.controllerAPI.getDefaultCloudTag((tag, err) => {console.log('tag:', tag, '| err:', err)})
// --> tag: cloud-demonstration | err: null;
app.controllerAPI.listClouds(console.log)
// --> check a single demonstration cloud is returned, with a localhost region and a single "empty" auth type;
app.controllerAPI.getClouds([], console.log)
// --> no results in clouds;
app.controllerAPI.getClouds(['cloud-no-such', 'cloud-demonstration'], console.log)
// --> proper results for cloud-demonstration, an err for cloud-no-such: "cloud cloud-no-such not found";


// In gijoe.

app.controllerAPI.getDefaultCloudTag((tag, err) => {console.log('tag:', tag, '| err:', err)})
// --> tag: cloud-$WHATEVER-YOU-NAMED-IT | err: null;
app.controllerAPI.listClouds(console.log)
// --> check available clouds are properly returned, based on your current provider type, for instance check on gce and lxd;
app.controllerAPI.getClouds(['cloud-no-such', 'cloud-$WHATEVER-YOU-NAMED-IT'], console.log)
// --> proper results for cloud-$WHATEVER-YOU-NAMED-IT, an err for cloud-no-such: "cloud no-such not found";
app.controllerAPI.getClouds([], console.log)
// --> no results in clouds;
```
Thank you!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5696/
Test PASSed.

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 1]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
           }, {immediate: true});
     });
 
+    describe('Cloud facade', function() {
+
+      it('can list available clouds', function(done) {
+        client.onmessage = function(received) {
+          const data = Y.JSON.parse(received.data);
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: No YUI please

`JSON.parse(received.data)`

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 2]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+          'request-id': 42,
+          type: 'Cloud',
+          request: 'Clouds',
+          params: {}
+        }));
+      });
+
+      it('can retrieves clouds by tag', function(done) {
+        client.onmessage = function(received) {
+          const data = Y.JSON.parse(received.data);
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: `JSON.parse(received.data)`

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 3]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+          params: {entities: [
+            {tag: 'cloud-no-such'},
+            {tag: 'cloud-demonstration'},
+          ]}
+        }));
+      });
+
+      it('can returns no results for no tags', function(done) {
+        client.onmessage = function(received) {
+          const data = Y.JSON.parse(received.data);
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: `JSON.parse(received.data)`

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 4]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+          assert.deepEqual(data, {
+            'request-id': 47,
+            response: {results: []}
+          });
+          done();
+        };
+        client.open();
+        client.send(Y.JSON.stringify({
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: `JSON.stringify(...`

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 5]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+          'request-id': 47,
+          type: 'Cloud',
+          request: 'Cloud',
+          params: {entities: []}
+        }));
+      });
+
+      it('can return the default cloud tag', function(done) {
+        client.onmessage = function(received) {
+          const data = Y.JSON.parse(received.data);
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: `JSON.parse(received.data)`

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 6]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+          assert.deepEqual(data, {
+            'request-id': 1,
+            response: {result: 'cloud-demonstration'}
+          });
+          done();
+        };
+        client.open();
+        client.send(Y.JSON.stringify({
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: `JSON.stringify(...`

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 7]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+                  regions: [{name: 'localhost'}],
+                  type: 'demonstration'
+                }
+              }]
+            }
+          });
+          done();
+        };
+        client.open();
+        client.send(Y.JSON.stringify({
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: `JSON.stringify(...`

File: jujugui/static/gui/src/test/test_sandbox_go.js [hunk 8]
```diff
@@ -1374,6 +1374,103 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+                  regions: [{name: 'localhost'}],
+                  type: 'demonstration'
+                }
+              }
+            }
+          });
+          done();
+        };
+        client.open();
+        client.send(Y.JSON.stringify({
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: `JSON.stringify(...`

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done.

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done.

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 3 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done.

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 4 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done.

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 5 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done.

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 6 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done.

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 7 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done.

File: jujugui/static/gui/src/test/test_sandbox_go.js (hunk 8 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_sandbox_go.js]: Done. and also done in all other occurrences in this file.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5698/
Test PASSed.

**Comment by hatched** [comment]
Thanks for adding these @frankban 

I ran the above commands in both deploys and everything worked as advertised.

As discussed it would be nice if we would stick to a callback argument convention. I propose the error-first callback convention from nodejs but that can be done as a follow-up

**Comment by makyo** [comment]
Thanks Francesco :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**